File: packages/primadb-python/scripts/mesh_agent.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import time

from primadb import Primadb


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--action", default="live", choices=["live", "verify-stored"])
    parser.add_argument("--relay", default="ws://127.0.0.1:9010")
    parser.add_argument("--ice-server", action="append", default=[])
    parser.add_argument("--room", default="primadb-mesh-agent")
    parser.add_argument("--replica", default=f"python-agent-{os.getpid()}")
    parser.add_argument("--storage-dir")
    parser.add_argument("--write-title")
    parser.add_argument("--write-body", default="python mesh agent")
    parser.add_argument("--expect-titles", default="")
    parser.add_argument("--min-peers", type=int, default=1)
    parser.add_argument("--timeout-ms", type=int, default=60_000)
    parser.add_argument("--hold-ms", type=int, default=2_000)
    parser.add_argument("--write-delay-ms", type=int, default=0)
    return parser.parse_args()


def parse_ice_server_specs(specs: list[str]) -> list[dict[str, object]]:
    servers: list[dict[str, object]] = []
    for spec in specs:
        trimmed = spec.strip()
        if trimmed.startswith("{"):
            value = json.loads(trimmed)
            if not isinstance(value, dict):
                raise ValueError("--ice-server JSON must decode to an object")
            servers.append(value)
        else:
            if not trimmed.startswith(("stun:", "turn:", "turns:")):
                raise ValueError(
                    f"invalid --ice-server value `{trimmed}`; use a STUN/TURN URL or JSON object"
                )
            servers.append({"urls": trimmed})
    return servers


def test_ice_servers() -> list[dict[str, object]]:
    return [{"urls": "stun:stun.cloudflare.com:3478"}]


def wait(ms: int) -> None:
    time.sleep(ms / 1000)


def collect_titles(db: Primadb, room: str) -> list[str]:
    entries = db.chain("boards").field(room).field("notes").query(
        {
            "order": {"path": "updated_at", "direction": "asc"},
            "limit": 1000,
        }
    )
    return [entry["value"]["title"] for entry in entries if entry.get("value", {}).get("title")]


def wait_for(predicate, timeout_ms: int, message: str):
    deadline = time.time() + (timeout_ms / 1000)
    last_value = None
    while time.time() < deadline:
        value = predicate()
        if value:
            return value
        last_value = value
        wait(100)
    raise RuntimeError(f"{message}. Last value: {last_value!r}")


def main() -> None:
    options = parse_args()
    expected_titles = [title for title in options.expect_titles.split(",") if title]
    db = Primadb(options.replica)
    storage = None

    if options.storage_dir:
      os.makedirs(options.storage_dir, exist_ok=True)
      storage = db.open_durable_storage(
          {
              "kind": "segment_files",
              "directory": options.storage_dir,
          }
      )

    if options.action == "verify-stored":
        titles = collect_titles(db, options.room)
        missing = [title for title in expected_titles if title not in titles]
        if missing:
            raise RuntimeError(f"Stored data missing titles: {', '.join(missing)}")
        print(
            json.dumps(
                {
                    "action": options.action,
                    "replica": options.replica,
                    "storage": storage,
                    "storedTitles": titles,
                    "python_package_storage_confirmed": True,
                },
                indent=2,
            )
        )
        return

    mesh = db.connect_mesh(
        {
            "room": options.room,
            "relayUrl": options.relay,
            "retryIntervalMs": 500,
            "iceServers": (
                parse_ice_server_specs(options.ice_server)
                if options.ice_server
                else test_ice_servers()
            ),
        }
    )

    try:
        wait_for(
            lambda: mesh.open_peer_count() >= options.min_peers,
            options.timeout_ms,
            f"Timed out waiting for {options.min_peers} open mesh peers",
        )

        if options.write_title:
            if options.write_delay_ms > 0:
                wait(options.write_delay_ms)
            now = int(time.time() * 1000)
            db.chain("boards").field(options.room).field("notes").set(
                {
                    "title": options.write_title,
                    "body": options.write_body,
                    "done": False,
                    "archived": False,
                    "created_at": now,
                    "updated_at": now,
                }
            )
            mesh.flush_pending()

        titles = wait_for(
            lambda: (
                current := collect_titles(db, options.room),
                current if all(title in current for title in expected_titles) else None,
            )[1],
            options.timeout_ms,
            "Timed out waiting for expected mesh titles",
        )

        print(
            json.dumps(
                {
                    "action": options.action,
                    "replica": options.replica,
                    "storage": storage,
                    "relay": options.relay,
                    "room": options.room,
                    "peerId": mesh.peer_id(),
                    "signaling": mesh.signaling_mode(),
                    "relayConnected": mesh.relay_connected(),
                    "openPeerCount": mesh.open_peer_count(),
                    "titles": titles,
                    "python_package_mesh_agent_confirmed": True,
                },
                indent=2,
            )
        )
        if options.hold_ms > 0:
            wait(options.hold_ms)
    finally:
        mesh.close()


if __name__ == "__main__":
    main()

File: packages/primadb-node/scripts/mesh-agent.mjs
#!/usr/bin/env node
import { Primadb } from "../index.js";
import { mkdirSync } from "node:fs";
import process from "node:process";

function parseArgs(argv) {
  const options = {
    action: "live",
    relay: "ws://127.0.0.1:9010",
    iceServers: [],
    room: "primadb-mesh-agent",
    replica: `node-agent-${process.pid}`,
    storageDir: undefined,
    writeTitle: undefined,
    writeBody: "node mesh agent",
    expectTitles: [],
    minPeers: 1,
    timeoutMs: 60_000,
    holdMs: 2_000,
    writeDelayMs: 0,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    const next = () => argv[++index];
    switch (arg) {
      case "--action":
        options.action = next();
        break;
      case "--relay":
        options.relay = next();
        break;
      case "--ice-server":
        options.iceServers.push(parseIceServerSpec(next()));
        break;
      case "--room":
        options.room = next();
        break;
      case "--replica":
        options.replica = next();
        break;
      case "--storage-dir":
        options.storageDir = next();
        break;
      case "--write-title":
        options.writeTitle = next();
        break;
      case "--write-body":
        options.writeBody = next();
        break;
      case "--expect-titles":
        options.expectTitles = next()
          .split(",")
          .map((value) => value.trim())
          .filter(Boolean);
        break;
      case "--min-peers":
        options.minPeers = Number(next());
        break;
      case "--timeout-ms":
        options.timeoutMs = Number(next());
        break;
      case "--hold-ms":
        options.holdMs = Number(next());
        break;
      case "--write-delay-ms":
        options.writeDelayMs = Number(next());
        break;
      default:
        throw new Error(`Unknown argument: ${arg}`);
    }
  }

  return options;
}

function parseIceServerSpec(spec) {
  const trimmed = String(spec).trim();
  if (trimmed.startsWith("{")) {
    const value = JSON.parse(trimmed);
    if (!value || typeof value !== "object" || Array.isArray(value)) {
      throw new Error("--ice-server JSON must decode to an object");
    }
    return value;
  }
  if (trimmed.startsWith("stun:") || trimmed.startsWith("turn:") || trimmed.startsWith("turns:")) {
    return { urls: trimmed };
  }
  throw new Error(`invalid --ice-server value \`${trimmed}\`; use a STUN/TURN URL or JSON object`);
}

function testIceServers() {
  return [{ urls: "stun:stun.cloudflare.com:3478" }];
}

function wait(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function waitFor(predicate, timeoutMs, errorMessage) {
  const deadline = Date.now() + timeoutMs;
  let lastValue = null;
  while (Date.now() < deadline) {
    const value = await predicate();
    if (value) {
      return value;
    }
    lastValue = value;
    await wait(100);
  }
  throw new Error(`${errorMessage}. Last value: ${JSON.stringify(lastValue)}`);
}

function collectTitles(db, room) {
  return db
    .chain("boards")
    .field(room)
    .field("notes")
    .query({
      order: { path: "updated_at", direction: "asc" },
      limit: 1000,
    })
    .map((entry) => entry?.value?.title)
    .filter(Boolean);
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  const db = new Primadb(options.replica);
  let storage = null;
  let mesh = null;

  try {
    if (options.storageDir) {
      mkdirSync(options.storageDir, { recursive: true });
      storage = db.openDurableStorage({
        kind: "segment_files",
        directory: options.storageDir,
      });
    }

    if (options.action === "verify-stored") {
      const titles = collectTitles(db, options.room);
      const missing = options.expectTitles.filter((title) => !titles.includes(title));
      if (missing.length > 0) {
        throw new Error(`Stored data missing titles: ${missing.join(", ")}`);
      }
      console.log(JSON.stringify({
        action: options.action,
        replica: options.replica,
        storage,
        storedTitles: titles,
        node_package_storage_confirmed: true,
      }, null, 2));
      return;
    }

    mesh = await db.connectMesh({
      room: options.room,
      relayUrl: options.relay,
      retryIntervalMs: 500,
      iceServers: options.iceServers.length > 0 ? options.iceServers : testIceServers(),
    });

    await waitFor(
      async () => (await mesh.openPeerCount()) >= options.minPeers,
      options.timeoutMs,
      `Timed out waiting for ${options.minPeers} open mesh peers`,
    );

    if (options.writeTitle) {
      if (options.writeDelayMs > 0) {
        await wait(options.writeDelayMs);
      }
      const now = Date.now();
      db.chain("boards").field(options.room).field("notes").set({
        title: options.writeTitle,
        body: options.writeBody,
        done: false,
        archived: false,
        created_at: now,
        updated_at: now,
      });
      await mesh.flushPending();
    }

    const titles = await waitFor(
      async () => {
        const current = collectTitles(db, options.room);
        return options.expectTitles.every((title) => current.includes(title)) ? current : null;
      },
      options.timeoutMs,
      "Timed out waiting for expected mesh titles",
    );

    console.log(JSON.stringify({
      action: options.action,
      replica: options.replica,
      storage,
      relay: options.relay,
      room: options.room,
      peerId: mesh.peerId(),
      signaling: mesh.signalingMode(),
      relayConnected: mesh.relayConnected(),
      openPeerCount: await mesh.openPeerCount(),
      titles,
      node_package_mesh_agent_confirmed: true,
    }, null, 2));
    if (options.holdMs > 0) {
      await wait(options.holdMs);
    }
  } finally {
    if (mesh) {
      await mesh.close();
    }
  }
}

main().catch((error) => {
  console.error(error.stack || String(error));
  process.exitCode = 1;
});
